ukm-semantics/main/preprocessing/endpoints.md

```diff
@@ -0,0 +1,282 @@
+                        | "decode_u64"  [token]
+                        | "decode_str"  [token]
+                        | "empty"  [token]
+                        | "ukm"  [token]
+                        | "CallData"  [token]
+                        | "EVMC_BAD_JUMP_DESTINATION"  [token]
+                        | "EVMC_SUCCESS"  [token]
+
+    syntax StringOrError  ::= methodSignature(String, NormalizedFunctionParameterList)  [function, total]
+                            | signatureTypes(NormalizedFunctionParameterList)  [function, total]
+                            | signatureType(Type)  [function, total]
+
+    rule methodSignature(S:String, Ps:NormalizedFunctionParameterList)
+        => concat(concat(S +String "(", signatureTypes(Ps)), ")")
+
+    rule signatureTypes(.NormalizedFunctionParameterList) => ""
+    rule signatureTypes(_ : T:Type , .NormalizedFunctionParameterList)
+        => signatureType(T)
+    rule signatureTypes
+            ( _ : T:Type
+            , ((_:NormalizedFunctionParameter , _:NormalizedFunctionParameterList)
+                #as Ps:NormalizedFunctionParameterList)
+            )
+        => concat(signatureType(T), concat(",", signatureTypes(Ps)))
+
+    rule signatureType(u8) => "Uint8"
+    rule signatureType(u16) => "Uint16"
+    rule signatureType(u32) => "Uint32"
+    rule signatureType(u64) => "Uint64"
+    rule signatureType(T) => error("Unknown type in signatureType:", ListItem(T))
+        [owise]
+
+    syntax Statement ::= signatureToCall
+                                    ( signature: Identifier
+                                    , signatures: List
+                                    , signatureToMethod: Map
+                                    , bufferId: Identifier
+                                    , gas: Identifier
+                                    ) [function, total]
+    rule signatureToCall
+            (... signature: _:Identifier
+            , signatures: .List
+            , signatureToMethod: _:Map
+            , bufferId: _:Identifier
+            , gas: _:Identifier
+            )
+        // TODO: Is this the right kind of error?
+        => :: state_hooks :: setStatus(:: ukm :: EVMC_BAD_JUMP_DESTINATION , .CallParamsList);
+
+    rule signatureToCall
+            (... signature: Signature:Identifier
+            , signatures: ListItem(CurrentSignature:String) Signatures:List
+            , signatureToMethod: (CurrentSignature |-> Method:Identifier _:Map) #as SignatureToMethod:Map
+            , bufferId: BufferId:Identifier
+            , gas: Gas:Identifier
+            )
+        =>  if signature == CurrentSignature {
+                .InnerAttributes
+                self . Method ( BufferId , Gas , .CallParamsList );
+                .NonEmptyStatements
+            } else {
+                .InnerAttributes
+                signatureToCall(Signature, Signatures, SignatureToMethod, BufferId, Gas)
+                .NonEmptyStatements
+            };
```

Comment:
I may be missing something here, but wouldn't it be easier just to have a `require` clause specifying that `signature` is present in `signatures`? This would simplify this rule and the one above it as well.

Reply:
You're right, I could have done something simpler, something like something like:
```
   syntax Expression ::= chooseSignature(Expression, Map, Expression, Expression)  [seqstrict(1,3,4)]
   rule chooseSignature(ptrValue(_, V:Value), SignatureToMethod, BufferId, Gas)
        => self . SignatureToMethod[Signature] (BufferId , Gas , .CallParamsList)
     requires Value in_keys(SignatureToMethod)
   rule chooseSignature(ptrValue(_, V:Value), SignatureToMethod, _, _) => :: state_hooks :: setStatus(:: ukm :: EVMC_BAD_JUMP_DESTINATION , .CallParamsList)
     requires notBool (Value in_keys(SignatureToMethod))

```

However, we are pretending that the "ukm::contract" macro/attribute is generating some contract struct + implementations, so it seemed more appropriate to generate exactly the same code in K.

Even so, maybe I should have went with the simpler option, given that we have a deadline, and I will do that if you think it's better. The difference between the two options did not seem to be that big to me, so I picked the one that seemed "cleaner", but that may not be the best choice.

So, should I switch to the simpler option, or should I leave the code as-is?

Reply:
I don't have any issues with leaving the code as-is, especially if, in the future, we move to a different implementation closer to the actual behavior of contract data generation.